associated-token-account/program/src/processor.rs

```diff
@@ -154,3 +160,132 @@ fn process_create_associated_token_account(
         ],
     )
 }
+
+/// Processes `CloseNested` instruction
+pub fn process_close_nested(program_id: &Pubkey, accounts: &[AccountInfo]) -> ProgramResult {
+    let account_info_iter = &mut accounts.iter();
+
+    let nested_associated_token_account_info = next_account_info(account_info_iter)?;
+    let nested_token_mint_info = next_account_info(account_info_iter)?;
+    let destination_token_account_info = next_account_info(account_info_iter)?;
+    let destination_system_account_info = next_account_info(account_info_iter)?;
+    let owner_associated_token_account_info = next_account_info(account_info_iter)?;
+    let owner_token_mint_info = next_account_info(account_info_iter)?;
+    let wallet_account_info = next_account_info(account_info_iter)?;
+    let spl_token_program_info = next_account_info(account_info_iter)?;
+    let spl_token_program_id = spl_token_program_info.key;
+
+    // Check owner address derivation
+    let (owner_associated_token_address, bump_seed) =
+        get_associated_token_address_and_bump_seed_internal(
+            wallet_account_info.key,
+            owner_token_mint_info.key,
+            program_id,
+            spl_token_program_id,
+        );
+    if owner_associated_token_address != *owner_associated_token_account_info.key {
+        msg!("Error: Owner associated address does not match seed derivation");
+        return Err(ProgramError::InvalidSeeds);
+    }
+
+    // Check nested address derivation
+    let nested_associated_token_address = get_associated_token_address_with_program_id(
+        owner_associated_token_account_info.key,
+        nested_token_mint_info.key,
+        spl_token_program_id,
+    );
+    if nested_associated_token_address != *nested_associated_token_account_info.key {
+        msg!("Error: Nested associated address does not match seed derivation");
+        return Err(ProgramError::InvalidSeeds);
+    }
+
+    let owner_associated_token_account_signer_seeds: &[&[_]] = &[
+        &wallet_account_info.key.to_bytes(),
+        &spl_token_program_id.to_bytes(),
+        &owner_token_mint_info.key.to_bytes(),
+        &[bump_seed],
+    ];
+
+    if !wallet_account_info.is_signer {
```

Comment:
i wonder all of victims for this nested ata still has their (possibly old) private keys... ;) Hope they hadn't burnt it with anger and agony. iirc, i advised some the nested token funds are irrecoverable several times.

btw, how are we going to provide the finding tool (or not?).

If CloseNested removes the explicit `destination_token_account_info`/`destination_system_account_info` flexibility, i think CloseNested can be permissionless like `ATA::Create`? (so, I'm inline-commenting here... lol)

Then, some hero can appear in the solana history, scanning the whole accounts and saving all. ;)

Anyway, this still can't save those who already lost his private keys.

Reply:
> btw, how are we going to provide the finding tool (or not?).

We could add it to the spl-token CLI, some sort of scan for accounts owned by your token accounts. Or even better, some of the offenders of using nested token accounts could provide this for their users? :wink: 

> If CloseNested removes the explicit destination_token_account_info/destination_system_account_info flexibility, i think CloseNested can be permissionless like ATA::Create? (so, I'm inline-commenting here... lol)

Very interesting idea... So anyone can recover funds for anyone else, since you're just moving from a nested ATA to a normal ATA owned by the wallet. I think I can get behind that. @t-nelson thoughts?

Reply:
Cool that we can do it.  I still hate all of this :slightly_smiling_face:

Reply:
During awful sleep in a plane, I realized that a permissionless version of this opens up an attack vector.

If anyone can close a nested ATA for someone, they can add an `InitializeAccount` instruction at the end of the transaction to resurrect the account under their name, and just hope that some dapp deposits funds into there.  It creates an easy way for malicious users to create trap accounts.

Reply:
> If anyone can close a nested ATA for someone, they can add an InitializeAccount instruction at the end of the transaction to resurrect the account under their name, and just hope that some dapp deposits funds into there. It creates an easy way for malicious users to create trap accounts.

hmm, it's true.. heh, it reminds of me _something_. ;) nice find. hmm, that's unfortunate. ;) maybe this can be prevented `realloc to 0 size` inside `Token::CloseAccount` on top of current zero-ing? (i know `relloc` isn't thing in mnb; but soon this time really)